### ISSUE TITLE
Parallel HPO with XGBoost GPU on EKS

### ISSUE THREAD

**Issue by zronaghi** [rapidsai/cloud-ml-examples#123]
https://github.com/rapidsai/cloud-ml-examples/issues/123#tasklist-block-d7d9e5ee-8b48-4404-83b6-44196945991f

**Comment by hcho3** [comment]
- [ ] Set up k8s cluster in AWS EKS
- [ ] Install Kubeflow
- [ ] Set up RAPIDS on Kubeflow
- [ ] Create Dask KubeCluster client
- [ ] Run Hyperparameter Optimization (HPO)

**Comment by hcho3** [comment]
Setting up EKS cluster:
```
eksctl create cluster --name phcho-first-cluster --region us-west-2 \
  --node-type g4dn.2xlarge --nodes 3 --ssh-access --ssh-public-key [key]
```

**Comment by hcho3** [comment]
Tip: the EKS web console may not show the cluster resources with error
```
Your current user or role does not have access to Kubernetes objects on this EKS cluster
```
Fix this by adding the correct IAM user or role with command
```
kubectl edit configmap aws-auth -n kube-system
```
https://stackoverflow.com/questions/70787520/your-current-user-or-role-does-not-have-access-to-kubernetes-objects-on-this-eks

**Comment by hcho3** [comment]
To install Kubeflow, use AWS's distribution: https://awslabs.github.io/kubeflow-manifests/docs/deployment/vanilla/guide/

**Comment by jacobtomlinson** [comment]
Thanks for figuring this out!

It's amazing how installing their distribution is `make` calling a Python script calling helm/kustomize many times. It really shows how far the Kubernetes packaging ecosystem has to go. I'd love to find out why it isn't a single `helm install` or `kubectl apply` command.

It also looks like their distribution is mostly the upstream vanilla resources with a couple of extra telemetry and sagemaker things thrown in.

Does it feel any different to the vanilla version you get from the `kubeflow/manifests` repo?

**Comment by hcho3** [comment]
I was not able to get the vanilla version working. It kept getting the error
```
no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1beata1"
```

But when I used the AWS's distribution, Kubeflow installed successfully.

**Comment by jacobtomlinson** [comment]
That's definitely a typo in `v1beata1`. I tend to check out a pinned version of the vanilla manifests, right now I'm testing with `1.5.1`. But using the AWS distribution on EKS is definitely the right choice 🙂.

**Comment by hcho3** [comment]
Update: I managed to set up `KubeCluster`, but only one of the two workers I requested connected to the scheduler. The other worker produced no log.
```
$ kubectl get pods --namespace kubeflow-user-example-com
NAME                                               READY   STATUS    RESTARTS   AGE
...
rapids-dask-default-worker-30239efd36              0/1     Pending   0          6m43s
rapids-dask-default-worker-46936f06df              1/1     Running   0          6m45s
rapids-dask-scheduler                              1/1     Running   0          6m48s

$ kubectl logs rapids-dask-scheduler --namespace kubeflow-user-example-com
...
2022-11-09 01:30:13,688 - distributed.scheduler - INFO - -----------------------------------------------
2022-11-09 01:30:14,301 - distributed.scheduler - INFO - State start
2022-11-09 01:30:14,309 - distributed.scheduler - INFO - -----------------------------------------------
2022-11-09 01:30:14,309 - distributed.scheduler - INFO - Clear task state
2022-11-09 01:30:14,310 - distributed.scheduler - INFO -   Scheduler at: tcp://192.168.74.182:8786
2022-11-09 01:30:14,310 - distributed.scheduler - INFO -   dashboard at:                     :8787
2022-11-09 01:30:20,525 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://192.168.54.247:40937', name: rapids-dask-default-worker-46936f06df, status: init, memory: 0, processing: 0>
2022-11-09 01:30:20,527 - distributed.scheduler - INFO - Starting worker compute stream, tcp://192.168.54.247:40937
2022-11-09 01:30:20,527 - distributed.core - INFO - Starting established connection

$ kubectl logs rapids-dask-default-worker-46936f06df --namespace kubeflow-user-example-com
...
2022-11-09 01:30:18,459 - distributed.nanny - INFO -         Start Nanny at: 'tcp://192.168.54.247:34195'
2022-11-09 01:30:19,938 - distributed.preloading - INFO - Creating preload: dask_cuda.initialize
2022-11-09 01:30:19,938 - distributed.preloading - INFO - Import preload module: dask_cuda.initialize
2022-11-09 01:30:20,130 - distributed.preloading - INFO - Run preload setup: dask_cuda.initialize
2022-11-09 01:30:20,503 - distributed.worker - INFO -       Start worker at: tcp://192.168.54.247:40937
2022-11-09 01:30:20,503 - distributed.worker - INFO -          Listening to: tcp://192.168.54.247:40937
2022-11-09 01:30:20,503 - distributed.worker - INFO -          dashboard at:       192.168.54.247:35361
2022-11-09 01:30:20,503 - distributed.worker - INFO - Waiting to connect to: tcp://rapids-dask-scheduler.kubeflow-user-example-com.svc.cluster.local:8786
2022-11-09 01:30:20,503 - distributed.worker - INFO - -------------------------------------------------
2022-11-09 01:30:20,503 - distributed.worker - INFO -               Threads:                          1
2022-11-09 01:30:20,503 - distributed.worker - INFO -                Memory:                  30.91 GiB
2022-11-09 01:30:20,503 - distributed.worker - INFO -       Local Directory: /tmp/dask-worker-space/worker-i4f0rq77
2022-11-09 01:30:20,504 - distributed.worker - INFO - Starting Worker plugin RMMSetup-89a0eab3-6c4c-4b5c-b119-ec7b922f3cdf
2022-11-09 01:30:20,504 - distributed.worker - INFO - Starting Worker plugin PreImport-a490db04-9e63-4a09-82dc-dcf77f72a164
2022-11-09 01:30:20,504 - distributed.worker - INFO - Starting Worker plugin CPUAffinity-027d3502-781e-4dc6-ae62-6d9a8110e155
2022-11-09 01:30:20,504 - distributed.worker - INFO - -------------------------------------------------
2022-11-09 01:30:20,527 - distributed.worker - INFO -         Registered to: tcp://rapids-dask-scheduler.kubeflow-user-example-com.svc.cluster.local:8786
2022-11-09 01:30:20,527 - distributed.worker - INFO - -------------------------------------------------
2022-11-09 01:30:20,528 - distributed.core - INFO - Starting established connection

$ kubectl logs rapids-dask-default-worker-30239efd36 --namespace kubeflow-user-example-com
[prints nothing!!]
```

**Comment by hcho3** [comment]
The HPO job appears to hang with no output.
```python
import optuna
import joblib
import dask.distributed

with dask.distributed.Client(cluster) as client:
    study = optuna.create_study(storage=f"mysql://root@{ip_address}:3306/metadb")
    # Optimize in parallel on your Dask cluster
    with joblib.parallel_backend("dask", client=client):
        study.optimize(objective, n_trials=100, n_jobs=-1)
    print(f"best_params = {study.best_params}")
```

Possible causes:
* The MySQL database is inaccessible to the workers.
* The workers crashes or disconnected.

**Comment by hcho3** [comment]
The MySQL database is not reachable. Running
```
mysql -u root -h 192.168.71.167 metadb
```
in the notebook instance leads to hang.

**Comment by hcho3** [comment]
Last time I tried the Dask Optuna, I couldn't get it to work due to missing packages in the workers. (It errors out when unpickling the objective function.) Any good ways to install packages on the worker side?

The MySQL server is running as part of the Kubeflow installation.

**Comment by jacobtomlinson** [comment]
You can install packages on the workers like this.

```python
from dask.distributed import PipInstall
plugin = PipInstall(packages=["git+https://github.com/jrbourbeau/optuna.git@dask-storage#egg=optuna"])
client.register_worker_plugin(plugin)
```

Or by setting the `EXTRA_PIP_PACKAGES` environment variable when you launch the cluster.

```python
cluster = KubeCluster(name="rapids-dask",
                      image=rapids_image,
                      worker_command="dask-cuda-worker",
                      n_workers=n_workers,
                      resources={"limits": {"nvidia.com/gpu": "1"}},
                      env={"DISABLE_JUPYTER": "true", "EXTRA_PIP_PACKAGES": "git+https://github.com/jrbourbeau/optuna.git@dask-storage#egg=optuna"})
```

**Comment by jacobtomlinson** [comment]
> The MySQL server is running as part of the Kubeflow installation.

I wonder if Istio is blocking this. Can you use the service name for the MySQL server instead of the IP?

**Comment by hcho3** [comment]
Logs from the pending worker:
```
$ kubectl describe pod rapids-dask-default-worker-5da4f6286a --namespace kubeflow-user-example-com
Name:           rapids-dask-default-worker-5da4f6286a
Namespace:      kubeflow-user-example-com
Priority:       0
Node:           <none>
Labels:         dask.org/cluster-name=rapids-dask
                dask.org/component=worker
                dask.org/workergroup-name=rapids-dask-default
                sidecar.istio.io/inject=false
Annotations:    kubernetes.io/psp: eks.privileged
Status:         Pending
...
  Warning  FailedScheduling  51s (x7 over 7m)  default-scheduler  0/3 nodes are available: 3 Insufficient nvidia.com/gpu.
```

**Comment by hcho3** [comment]
Update: I was able to run the notebook [xgboost-gpu-hpo-job-parallel-k8s.ipynb](https://github.com/rapidsai/cloud-ml-examples/blob/f8dab3e78d0fe825c22717be47d51ad89b8c8cf2/gke-dask/notebooks/xgboost-gpu-hpo-job-parallel-k8s.ipynb) end to end successfully using a GKE cluster. The only issue is that the Dask cluster had access to only a single worker. (See my previous comment.)

Also, I managed to use the vanilla manifest to install Kubeflow. Unlike the AWS's distribution, the vanilla manifest is much simpler to use and deploy. Here's how:

1. Create an EKS cluster using the AWS CLI:
```
$ eksctl create cluster --name phcho-first-cluster  --version 1.22 --region us-west-2 --nodegroup-name linux-nodes \
  --node-type g4dn.2xlarge --nodes 3 --managed
```

2. [Install the latest Kustomize](https://kubectl.docs.kubernetes.io/installation/kustomize/)
```
$ kustomize version
{Version:kustomize/v4.5.5 GitCommit:daa3e5e2c2d3a4b8c94021a7384bfb06734bcd26 BuildDate:2022-05-20T20:25:40Z GoOs:windows GoArch:amd64}
```

3. Check out and build the vanilla manifest.
```
$ git clone https://github.com/kubeflow/manifests.git -b v1.6.1
$ cd manifests
$ kustomize build example > example.yaml
```

4. Apply the manifest to the EKS cluster. Note that you may need to repeat the command a few times in order for it to complete successfully.
```
$ kubectl apply -f example.yaml   # repeat a few times until successful
```

5. Install the Dask K8s operator using Helm:
```
helm install --repo https://helm.dask.org --create-namespace -n dask-operator \
  --generate-name dask-kubernetes-operator
```

6. Grant necessary permissions to the notebook instance to manage custom Dask resources.
```
$ kubectl edit clusterrole kubeflow-kubernetes-edit
...
rules:
...
- apiGroups:
  - "kubernetes.dask.org"
  verbs:
  - "*"
  resources:
  - "*"
...
```

**Comment by hcho3** [comment]
From doing more experiments, it appears that we have to allocate `N + 2` EC2 instances to the EKS clusters, where `N` is the desired number of Dask workers.

So for example, to create a Dask cluster with 4 workers, we need an EKS cluster with 6 instances.
* 1 instance hosts the Dask scheduler
* 4 instances host Dask workers
* 1 instance doesn't host any Dask-related pod.
